src/chainstate/stacks/index/storage.rs

```diff
@@ -919,154 +646,133 @@ impl TrieFileStorage {
-
-        // create an empty file for this block, so we can't extend to it again
-        let block_dir = TrieFileStorage::block_dir(&self.dir_path, bhh);
-        let block_path = TrieFileStorage::block_path(&self.dir_path, bhh);
-        let block_path_tmp = TrieFileStorage::block_path_tmp(&self.dir_path, bhh);
-        // check if block_path exists
-        match fs::metadata(&block_path) {
-            Ok(_) => {
-                trace!("Block path exists: {:?}", &block_path);
-                return Err(Error::ExistsError);
-            },
-            Err(e) => {
-                if e.kind() != io::ErrorKind::NotFound {
-                    return Err(Error::IOError(e));
-                }
-            }
-        };
-        // check if block_path_tmp exists
-        match fs::metadata(&block_path_tmp) {
-            Ok(_) => {
-                error!("Tried to create index block {:?} twice", bhh);
-                return Err(Error::ExistsError);
-            },
-            Err(e) => {
-                if e.kind() != io::ErrorKind::NotFound {
-                    return Err(Error::IOError(e));
-                }
-            }
-        };
-
-        fs::create_dir_all(block_dir)
-            .map_err(Error::IOError)?;
+        trie_sql::add_chain_tip(&self.db, bhh)?;
 
-        trace!("Extend from {:?} to {:?} in {:?}", &self.cur_block, bhh, &block_path_tmp);
+        let trie_buf = TrieRAM::new(bhh, size_hint, &self.cur_block);
 
-        // write the new file out and add its parent
-        let mut fd = fs::OpenOptions::new()
-            .read(true)
-            .write(true)
-            .create_new(true)
-            .open(&block_path_tmp)
-            .map_err(|e| {
-                if e.kind() == io::ErrorKind::NotFound {
-                    error!("File not found: {:?}", &block_path_tmp);
-                    Error::NotFoundError
-                }
-                else {
-                    Error::IOError(e)
-                }
-            })?;
+        // place a lock on this block, so we can't extend to it again
+        if !trie_sql::lock_bhh_for_extension(&mut self.db, bhh)? {
+            warn!("Block already extended: {}", &bhh);
+            return Err(Error::ExistsError);
+        }
 
-        TrieRAM::write_trie_file(&mut fd, &[], &[], identifier, bhh)?;
+        trace!("Extended from {} to {}", &self.cur_block, bhh);
 
-        fd.write_all(self.cur_block.as_bytes())
-            .map_err(|e| Error::IOError(e))?;
+        // write the new file out and add its parent
+        // TODO: I don't think this is necessary in the SQL impl,
+        //   so I've commented out -- but if it turns out we need to write
+        //   the parent block somewhere at this point, I've left this todo here so as not
+        //   to forget.
+        //  TrieRAM::write_trie_file(&mut fd, &[], &[], identifier, bhh)?;
+        //  fd.write_all(self.cur_block.as_bytes())
+        //    .map_err(|e| Error::IOError(e))?;
```

Comment:
I'm pretty sure the reason for this was to implement the `Error::ExistsError` check above, but we'll want to wait for tests to confirm.

Reply:
Looks like we're good on this.